refactor!: prqlc-parser major reorg changes, remove prqlc-ast

m-span opened the pull request:
From #4603 , we have the following new terminology:

| stage	|sub-stage	|AST
|----|----|----|
 |parse	|lexer	|string ->  _LR — Lexer Representation_	
 |parse	|parser	|tokens -> _PR — Parser Representation_	
 |semantic|	ast_expand	|PR -> _PL — Pipelined Language_	
 |semantic	|resolver	|PL	
 |semantic	|flatten	|PL	
 |semantic	|lowering	|PL -> _RQ — Resolved Query_	
 |sql	|preprocess	|RQ	
 |sql	|srq-compiler	|RQ ->  _PQ — Partitioned Query_	
 |sql	|postprocess	|PQ
 |sql	|sql-compiler	|PQ -> sqlparser::ast
 |sql|	codegen|	sqlparser::ast -> string


Goals:

- [x] Move all files from `prqlc-ast` to `prqlc-parser`. `prqlc-ast` is to be deprecated
- [x] Within parser stage, rename objects, modules and methods to meet new spec. Organize lexer and parser stage into separate modules.

m-span committed:
prqlc-parser refactor

max-sixty commented on prqlc/prqlc-parser/src/lib.rs:
```diff
@@ -1,35 +1,25 @@
+pub mod span;
+
+mod test;
+// pub use crate::err::*;
```
Super minor and let's not delay merging — but IIUC the rust standard is putting `mod`s above `use`s; the PR seems to reverse those.

(another one where I don't have a view apart from doing the standard thing — so if I'm wrong on the standard thing then let's do the correct standard thing :) )

max-sixty replied:
I made some changes for these

m-span committed:
ok that somehow fixed some of the issues

max-sixty committed:
Remove `ast_code_matches` file

No longer needed; we're not going to have multiple copies of AST items any longer...

max-sixty committed:
remove all traces of `prqlc-ast` :)

max-sixty committed:
some cleanups

max-sixty committed:
Run import sorting

max-sixty commented on prqlc/prqlc-parser/src/parser/pr/mod.rs:
```diff
@@ -0,0 +1,21 @@
+//this exports everything within as if it was at the root.
+//eg: PR::expr::Expr becomes PR::Expr
+pub use expr::*;
+pub use ident::*;
+pub use ops::*;
+pub use stmt::*;
+pub use types::*;
```
This is pretty aggressive exporting. No need to delay merge, but was this the case before? If not, would be good to trim it down at some point...

max-sixty replied:
OK it was like this prior! Closing

m-span replied:
yep, another place for export optimization. Saving that for the next 'pass'

max-sixty committed:
mod vs use; small refinements

max-sixty commented on prqlc/prqlc-parser/src/parser/pr/mod.rs:
Not for this PR (this existed [before](https://github.com/PRQL/prql/pull/4634/files#diff-77aa79b7ef4fed0c42b75da58df055402450091a93c1ca7a8b5d96f92c34e5bc) — but exporting `foo::*` _and_ exporting the module `foo` seems prone to confusion and IIUC serves no purpose — guarantees there are two ways of accessing the same variable (I think?)

m-span replied:
My first stab at this code involved rewriting all the imports and exports. However, this made the refactor twice as hard, so I tried to export things in a way that `prqlc-compiler` is expecting. That said, we could definitely optimize things.

One such question, for example, is `lexer::lr::Literal`. It is used in every subsequent AST. Right now, they are (almost) all referencing `lexer::lr::Literal`.
The other option is that each AST re-exports all the objects it uses. So, even though only one `Literal` exists, there will also be a `pr::Literal`, `pl::Literal,` `q::literal`

m-span replied:
PR will reference `lr::Literal`... PL will only have to import PR and its `pr::Literal`, etc etc

max-sixty replied:
Yes good idea to push into another PR.

No strong view on exactly how that's done. I think the existing state is strictly worse than some other alternatives since it's duplicating the locations of everything without providing convenience, so definitely some improvements possible.

max-sixty committed:
add comment re exports

max-sixty commented on prqlc/prqlc-parser/src/error/mod.rs:
```diff
@@ -227,121 +229,3 @@ impl<T, E: WithErrorInfo> WithErrorInfo for Result<T, E> {
 }
-
-#[cfg(test)]
-mod tests {
```
@m-span did we lose this? A quick ctrl-f doesn't get any hits, am I missing something?

m-span replied:
i did some other minor refactors in parser while restructuring. Large file with tests -> tests in separate file

max-sixty replied:
Sorry, GitHub wasn't showing the diff, found it when searching locally

max-sixty commented on prqlc/prqlc-parser/src/lexer/mod.rs:
```diff
@@ -428,195 +473,3 @@ fn end_expr() -> impl Parser<char, (), Error = Cheap<char>> {
 pub struct TokenVec(pub Vec<Token>);
-
-#[cfg(test)]
-mod test {
```
(checked this and this moved to `test.rs`)

m-span replied:
👍

max-sixty committed:
Use attributes on `mod` rather than within the test file

max-sixty committed:
Meh, no need for `mod` as the only item in a module

max-sixty commented on prqlc/prqlc-parser/src/lexer/lr/mod.rs:
Same here — IIUC it would be strictly better to have only one of these — no strong view on which

m-span committed:
changelog update

m-span committed:
Merge branch 'PRQL:main' into renaming

max-sixty commented on CHANGELOG.md:
```diff
@@ -20,6 +20,11 @@
+- Major reorganization of `prqlc-parser`. Files have been re-arranged and
+  exports have changed
+
+- Deprecated `prqlc-ast`. Functionality and objects were moved to `prqlc-parser`
```
```suggestion
- Major reorganization of `prqlc-parser` — `prqlc-ast` is merged into `prqlc-parser`, and `prqlc-parser`'s files are rearranged, including its exports. This is part of an effort to modularize the compiler by stage, reducing the amount of context that's required to understand a single stage. There will likely be some further changes (more detail in the PR description). (@m-span, #4634)
  -  This is a breaking change for any libraries that use `prqlc-parser` which should be fairly rare.
```

(I actually wiped it `prqlc-ast`; don't think there's any need to keep it around...)

max-sixty committed:
Update CHANGELOG.md

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

max-sixty commented on CHANGELOG.md:
```diff
@@ -20,6 +20,16 @@
+  (@m-span, #4634)
+
+  - This is a breaking change for any libraries that use `prqlc-parser` which
+    should be fairly rare.
```
```suggestion
  - This is a breaking change for any libraries that depend on `prqlc-parser` (which
    should be fairly rare).
```

max-sixty committed:
Update CHANGELOG.md

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
